bundles/org.openhab.core.crypto.cryptostore/src/main/java/org/openhab/core/crypto/cryptostore/CryptoStore.java

```diff
@@ -0,0 +1,328 @@
+    }
+
+    public void setKeystoreFileName(String keystoreFileName) {
+        this.keystoreFileName = keystoreFileName;
+    }
+
+    public String getKeystoreFileName() {
+        return this.keystoreFileName;
+    }
+
+    public void setKeys(String privKey, byte[] keyString, String cert) throws GeneralSecurityException, Exception {
+        setPrivKey(privKey, keyString);
+        setCert(cert);
+    }
+
+    public void setKeyStore(String keystoreFileName) {
+        this.keystoreFileName = keystoreFileName;
+    }
+
+    public void loadFromKeyStore(String keystoreFileName, String keystorePassword, byte[] keyString)
+            throws GeneralSecurityException, IOException, Exception {
+        this.keystoreFileName = keystoreFileName;
+        loadFromKeyStore(keystorePassword, keyString);
+    }
+
+    public void loadFromKeyStore(String keystorePassword, byte[] keyString)
+            throws GeneralSecurityException, IOException, Exception {
+        Key key = convertByteToKey(keyString);
+        KeyStore keystore = KeyStore.getInstance("JKS");
+        FileInputStream keystoreInputStream = new FileInputStream(this.keystoreFileName);
+        keystore.load(keystoreInputStream, keystorePassword.toCharArray());
+        byte[] byteKey = keystore.getKey(this.alias, keystorePassword.toCharArray()).getEncoded();
+        this.privKey = encrypt(new String(Base64.getEncoder().encode(byteKey)), key);
+        setCert(keystore.getCertificate(this.alias));
+        String caAlias = this.caAlias;
+        Certificate caCert = keystore.getCertificate(caAlias);
+        if (caCert != null) {
+            setCaCert(caCert);
+        }
+    }
+
+    public KeyStore getKeyStore(String keystorePassword, byte[] keyString)
+            throws GeneralSecurityException, IOException, Exception {
+        KeyStore keystore = KeyStore.getInstance("JKS");
+        keystore.load(null, null);
+        byte[] pkcs8EncodedBytes = Base64.getDecoder().decode(getPrivKey(keyString));
+        PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(pkcs8EncodedBytes);
+        KeyFactory kf = KeyFactory.getInstance(this.keystoreAlgorithm);
+        keystore.setKeyEntry(this.alias, kf.generatePrivate(keySpec), keystorePassword.toCharArray(),
+                new java.security.cert.Certificate[] { getCert() });
+        if (!caCert.isEmpty()) {
+            String caAlias = this.caAlias;
+            keystore.setCertificateEntry(caAlias, getCaCert());
+        }
+        return keystore;
+    }
+
+    public void saveKeyStore(String keystorePassword, byte[] keyString)
+            throws GeneralSecurityException, IOException, Exception {
+        saveKeyStore(this.keystoreFileName, keystorePassword, keyString);
+    }
+
+    public void saveKeyStore(String keystoreFileName, String keystorePassword, byte[] keyString)
+            throws GeneralSecurityException, IOException, Exception {
+        FileOutputStream keystoreStream = new FileOutputStream(keystoreFileName);
```

Comment:
If two bindings accidentally choose the same filename, they could overwrite each other's keystore.

Reply:
Unfortunately I'm not sure how to resolve that in this structure.  In theory, the binding review process should catch accidental use of another binding name (or something generic like my.keystore).

Reply:
Would a binding need more than one keystore? Otherwise it might be best to just use the binding id? (Together with my other suggestion above)

Reply:
Absolutely.  Even at the individual thing level yes.  For example, ATV uses a mechanism that creates a MITM for purposes of debugs.  I have to maintain a keystore for each side of that just for that connection.  Also, ATV works with two different protocol stacks in the case of ShieldTV, each uses it's own KeyStore.

Reply:
Then this would probably have to be solved with an instruction on how to name the file (e.g bindingId + identifier) that's in the javadoc so it can be (more) easily checked when doing reviews

Reply:
Instead of taking keystoreName as a parameter, what if we determine keystore name based on the namespace of the caller?  Bindings package names don't change, and I think it's pretty easy to walk up a stack trace tree to figure out who's invoking the method.  I can work on some sample code if it would help

Reply:
I went down the road of a single keystore and there were several issues.  First, working across things made it complex.  Second, the CA path was different for each which added complexity (because they were self signed certs by the device).  Ultimately, the work involved with making a single keystore wasn't worth it when maintaining multiple keystores was easy and kept a natural divide across the keys.